News channel message publish

MattTheDev opened the pull request:
## Summary

This PR adds in the ability to Publish a new message within a news chnanel. If the channel type is not a news channel, it'll ignore the request to publish. I thought about returning an error here, and this could easily be added. 

Resolves #1528 

## Changes

Addition of the API handler for /crosspost endpoint. 
Addition of PublishAsync to the Socket and Rest Messages.
Only executes if the channels are news channel types.

Commit:
Added PublishAsync to Messages.

Commit:
Added missing implementation.

SubZero0 commented on src/Discord.Net.Core/Entities/Messages/IUserMessage.cs:
```diff
@@ -57,6 +57,15 @@ public interface IUserMessage : IMessage
+        /// <returns>
+        ///     A task that represents the asynchronous operation for publishing this message.
+        /// </returns>
+        Task PublishAsync(RequestOptions options = null);
```
I believe we should stick to the endpoint name, so it would be `CrosspostAsync`

MattTheDev replied:
Aligned naming w/ endpoint.

SubZero0 commented on src/Discord.Net.Core/Entities/Messages/IUserMessage.cs:
```diff
@@ -57,6 +57,15 @@ public interface IUserMessage : IMessage
         Task UnpinAsync(RequestOptions options = null);
 
+        /// <summary>
+        ///     Publish this message when executed in a News Channel.
```
We could split this into `Publish (crosspost) this message.` and a `<remark>` notifying about it being only available for news channels.

MattTheDev replied:
Fixed sentence / added remarks.

SubZero0 commented on src/Discord.Net.Rest/DiscordRestApiClient.cs:
```diff
@@ -695,6 +695,15 @@ public async Task TriggerTypingIndicatorAsync(ulong channelId, RequestOptions op
             var ids = new BucketIds(channelId: channelId);
             await SendAsync("POST", () => $"channels/{channelId}/typing", ids, options: options).ConfigureAwait(false);
         }
+        public async Task PublishAsync(ulong channelId, ulong messageId, RequestOptions options = null)
```
Imo, this name should stick with the endpoint name, `CrosspostMessageAsync`

MattTheDev replied:
Aligned naming w/ endpoint.

SubZero0 commented on src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs:
```diff
@@ -279,5 +285,14 @@ public static MessageSource GetSource(Model msg)
+        public static Task PublishAsync(IMessage msg, BaseDiscordClient client, RequestOptions options)
+            => PublishAsync(msg.Channel.Id, msg.Id, client, options);
+
+        public static async Task PublishAsync(ulong channelId, ulong msgId, BaseDiscordClient client,
```
I would also follow the endpoint name for these.

MattTheDev replied:
Aligned naming w/ endpoint.

SubZero0 commented on src/Discord.Net.Rest/Entities/Messages/RestUserMessage.cs:
```diff
@@ -148,6 +148,14 @@ public Task ModifySuppressionAsync(bool suppressEmbeds, RequestOptions options =
 
+        public async Task PublishAsync(RequestOptions options = null)
+        {
+            if (Channel.GetType() == typeof(RestNewsChannel))
```
Don't think failing silently is a behavior that's accepted.
There should be a check and throw an exception.
The check could also be changed to `Channel is RestNewsChannel`.

MattTheDev replied:
Threw InvalidOperation - seemed the most logical.

SubZero0 commented on src/Discord.Net.Rest/Entities/Messages/RestUserMessage.cs:
```diff
@@ -148,6 +148,14 @@ public Task ModifySuppressionAsync(bool suppressEmbeds, RequestOptions options =
             TagHandling roleHandling = TagHandling.Name, TagHandling everyoneHandling = TagHandling.Ignore, TagHandling emojiHandling = TagHandling.Name)
             => MentionUtils.Resolve(this, 0, userHandling, channelHandling, roleHandling, everyoneHandling, emojiHandling);
 
+        public async Task PublishAsync(RequestOptions options = null)
```
Missing `/// <inheritdoc />`

MattTheDev replied:
Added.

SubZero0 commented on src/Discord.Net.WebSocket/Entities/Messages/SocketUserMessage.cs:
```diff
@@ -150,5 +150,13 @@ public Task ModifySuppressionAsync(bool suppressEmbeds, RequestOptions options =
         private string DebuggerDisplay => $"{Author}: {Content} ({Id}{(Attachments.Count > 0 ? $", {Attachments.Count} Attachments" : "")})";
         internal new SocketUserMessage Clone() => MemberwiseClone() as SocketUserMessage;
+
+        public async Task PublishAsync(RequestOptions options = null)
+        {
+            if (Channel.GetType() == typeof(SocketNewsChannel))
```
Same as above for `RestUserMessage` (and I would probably reallocate this for the sake of aesthetics :p)

MattTheDev replied:
Fixed and moved! :)

MattTheDev commented:
Great feedback, @SubZero0. Working on resolving them now.

Commit:
1. Aligned with naming standards
2. Clarified xml docs
3. Properly threw exceptions instead of failing silently.

Still34 commented on src/Discord.Net.Core/Entities/Messages/IUserMessage.cs:
```diff
@@ -57,6 +57,18 @@ public interface IUserMessage : IMessage
         Task UnpinAsync(RequestOptions options = null);
 
+        /// <summary>
+        ///     Publish (crosspost) this message.
```
Use third-person present tense for summary (i.e., `Publishes`/`crossposts`)

MattTheDev replied:
Corrected to standard tense.

Still34 commented on src/Discord.Net.Core/Entities/Messages/IUserMessage.cs:
```diff
@@ -57,6 +57,18 @@ public interface IUserMessage : IMessage
+        ///     A task that represents the asynchronous operation for publishing this message.
+        /// </returns>
+        /// <remarks>
+        ///     Publishing (crossposting) is only available in news channels.
```
Remarks should describe what the method does, not just what to look out for. A better way to phrase this is 

```xml
<remarks>
<note type="warning">
This method will throw an [insert exception xref here] if attempted to publish in a non-news channel.
</note>
This method will crosspost the message. Please note that publishing or crossposting is only available in news channels.
</remarks>
```

MattTheDev replied:
Sorry for wasting your time on this! Appreciate the guidance. Modified verbiage.

Still34 commented on src/Discord.Net.Rest/Entities/Messages/RestUserMessage.cs:
```diff
@@ -148,6 +148,17 @@ public Task ModifySuppressionAsync(bool suppressEmbeds, RequestOptions options =
+        {
+            if (!(Channel is RestNewsChannel))
+            {
+                throw new InvalidOperationException("Publishing (crossposting) is only valid in news channels.");
```
Please document the exception with `<exception>` tag. See [Sandcastle <exception> tag documentation](https://ewsoftware.github.io/XMLCommentsGuide/html/bbd1e65d-c87c-4b46-9a1a-259d3c5cd936.htm).

MattTheDev replied:
Resolved. And noted for next time.

Still34 commented on src/Discord.Net.WebSocket/Entities/Messages/SocketUserMessage.cs:
```diff
@@ -147,7 +147,18 @@ public Task ModifySuppressionAsync(bool suppressEmbeds, RequestOptions options =
+        {
+            if (!(Channel is SocketNewsChannel))
+            {
+                throw new InvalidOperationException("Publishing (crossposting) is only valid in news channels.");
```
Ditto above.

MattTheDev replied:
Thanks for the guidance / assistance. Tidied up all feedback.

Still34 reviewed (changes requested):
Docs strings need to be tidied up before merge.

MattTheDev committed:
Additional documentation included.

MattTheDev committed:
Removed un-needed comments.

MattTheDev reviewed (commented):
Modified per your request.
